### PR TITLE
Add default init function to object to hold fields

### LINKED ISSUE #15240: Object initialization does not follow the initialization protocol outlined in the spec
The initialization steps given in the spec are as follows:
- allocating storage for the object
- initializing each field with its default value, if it has one
- initializing the methods of the object using the type's method definitions
- calling the object's __init method, if there is one

Currently, the initialization of fields to their default values is also done by the __init() method. Hence, explicitly calling __init() results in default values getting set as well.

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
According to the previous implementation, the initialization of fields of an object is done inside the __init() function. Because of that, initialization happens each and every time we call the self.__init(). This PR fixes that issue. 

Fixes #15240

## Approach
Create another init function called default init function and initialization of fields are done inside the methods.

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRTypeWriter.java [hunk 1]
```diff
@@ -343,9 +344,17 @@ private void writeObjectAndServiceTypes(BObjectType bObjectType) {
             attachedFuncs = new ArrayList<>();
             buff.writeByte(0); // constructor not present
         }
-        buff.writeInt(attachedFuncs.size());
+
+        List<BAttachedFunction> writableAttachedFunctions = new ArrayList<>();
         for (BAttachedFunction attachedFunc : attachedFuncs) {
-            writeAttachFunction(attachedFunc);
+            if (bObjectType.getKind() == TypeKind.SERVICE && attachedFunc.funcName == Names.DEFAULT_INIT_SUFFIX) {
```
Comment (lankavitharana) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRTypeWriter.java]: shall we move this check to runtime?  as we have discussed, the proper fix would be to keep these as separate functions in the type, but for the time being +1 for moving this check to runtime

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 2]
```diff
@@ -3373,6 +3378,31 @@ private BLangStatementExpression desugarObjectTypeInit(BLangTypeInit typeInitExp
         BLangSimpleVariableDef objVarDef = createVarDef("$obj$", objType, typeInitExpr, typeInitExpr.pos);
         BLangSimpleVarRef objVarRef = ASTBuilderUtil.createVariableRef(typeInitExpr.pos, objVarDef.var.symbol);
         blockStmt.addStatement(objVarDef);
+
+        BInvokableSymbol defaultInitializerSymbol = null;
+        if (typeInitExpr.type.getKind() == TypeKind.UNION) {
```
Comment (lankavitharana) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: is this case valid? is it possible to get a union type here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRTypeWriter.java (hunk 1 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRTypeWriter.java]: This is changed in 94dc802.

File: tests/jballerina-unit-test/src/test/resources/test-src/object/ObjectProject/src/init/object-initializer.bal [hunk 3]
```diff
@@ -241,3 +241,42 @@ function testCheckPanicInObjectInitArg() returns error {
...
+        self.__init("B+");
+        return self.marks;
+    }
+};
+
+function testInitInvocationInsideObjectForChangingFieldValuesWithArgs() returns int {
+    Student2 student = new("B");
+    student.marks = 82;
+    return student.getMarks();
+}
```
Comment (KavinduZoysa) [inline tests/jballerina-unit-test/src/test/resources/test-src/object/ObjectProject/src/init/object-initializer.bal]: Add more test cases
1. with expression
2. verify user define init function

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 4]
```diff
@@ -347,18 +347,19 @@ private void addAttachedFunctionsToPackageLevel(BLangPackage pkgNode, SymbolEnv
                 if (objectTypeNode.flagSet.contains(Flag.ABSTRACT)) {
                     continue;
                 }
+                objectTypeNode.defaultInitFunction = createInitFunctionForObjectType(objectTypeNode, env);
```
Comment (KavinduZoysa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: change the name defaultInitFunction ->

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 5]
```diff
@@ -3373,6 +3378,31 @@ private BLangStatementExpression desugarObjectTypeInit(BLangTypeInit typeInitExp
         BLangSimpleVariableDef objVarDef = createVarDef("$obj$", objType, typeInitExpr, typeInitExpr.pos);
         BLangSimpleVarRef objVarRef = ASTBuilderUtil.createVariableRef(typeInitExpr.pos, objVarDef.var.symbol);
         blockStmt.addStatement(objVarDef);
+
+        BInvokableSymbol defaultInitializerSymbol = null;
```
Comment (KavinduZoysa) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: get the Symbol from objType

**Comment by KavinduZoysa** [comment]
The same PR(#20241) is sent to the master branch with suggested requests.

**Pull request closed without merging**